File: src/logger.py
from __future__ import print_function
from termcolor import colored
import datetime
import inspect
import os

terminal = {
    'normal': '\033[0m',
    'bright': '\033[1m',
    'invert': '\033[7m',
    'black': '\033[30m', 
    'red': '\033[31m', 
    'green':'\033[32m', 
    'yellow': '\033[33m', 
    'blue': '\033[34m', 
    'magenta': '\033[35m',
    'cyan': '\033[36m',
    'white': '\033[37m', 
    'default': '\033[39m'
}

def get(filename=None):
    return Logger(filename)

class Logger(object):

    def __init__(self, filename=None):
        now = datetime.datetime.now()
        self.verbose_thresh = os.environ.get('VERBOSE', 0)
        if filename is not None:
            self.filename = \
                '{}-{:04d}{:02d}{:02d}-{:02d}{:02d}{:02d}.log'.format(
                filename, 
                now.year, now.month, now.day, now.hour, now.minute, now.second)
            dirname = os.path.dirname(self.filename)
            if not os.path.exists(dirname):
                os.makedirs(dirname)
            open(self.filename, 'w').close()
        else:
            self.filename = None
        pass
    
    @staticmethod
    def get_time_str(t=datetime.datetime.now()):
        timestr = '{:04d}-{:02d}-{:02d} {:02d}:{:02d}:{:02d}'.format(
                t.year, t.month, t.day, t.hour, t.minute, t.second)
        return timestr

    def log(self, message, typ='info', verbose=0):
        if typ == 'info':
            typstr_print = '{0}INFO:{1}'.format(
                    terminal['green'], terminal['default'])
            typstr_log = 'INFO:'
        elif typ == 'warning':
            typstr_print = '{0}WARNING:{1}'.format(
                    terminal['yellow'], terminal['default'])
            typstr_log = 'WARNING'
        elif typ == 'error':
            typstr_print = '{0}ERROR:{1}'.format(
                    terminal['red'], terminal['default'])
            typstr_log = 'ERROR'
        elif typ == 'fatal':
            typstr_print = '{0}FATAL:{1}'.format(
                    terminal['red'], terminal['default'])
            typstr_log = 'FATAL'
        else:
            raise Exception('Unknown log type: {0}'.format(typ));
        timestr = self.get_time_str()
        for (frame, filename, line_number, function_name, lines, index) in \
                inspect.getouterframes(inspect.currentframe()):
            if not filename.endswith('logger.py'):
                break
        cwd = os.getcwd()
        if filename.startswith(cwd): filename = filename[len(cwd):]
        filename = filename.lstrip('/')
        callerstr = '{0}:{1}'.format(filename, line_number)
        printstr = '{0} {1} {2} {3}'.format(
                typstr_print, timestr, callerstr, message)
        logstr = '{0} {1} {2} {3}'.format(
                typstr_log, timestr, callerstr, message)
        if self.verbose_thresh <= verbose:
            print(printstr)
        if self.filename is not None:
            with open(self.filename, 'a') as f:
                f.write(logstr)
                f.write('\n')
        pass

    def info(self, message, verbose=0):
        self.log(message, typ='info', verbose=verbose)
        pass

    def warning(self, message, verbose=0):
        self.log(message, typ='warning', verbose=verbose)
        pass

    def error(self, message, verbose=0):
        self.log(message, typ='error', verbose=verbose)
        pass

    def fatal(self, message, verbose=0):
        self.log(message, typ='fatal', verbose=verbose)
        sys.exit(0)
        pass

File: src/utils/logger.lua
local os = require('os')
local torch = require('torch')
local path = require('utils.path')
local LoggerClass = torch.class('Logger')
local verboseThreshold = os.getenv('VERBOSE')
local logToEnv = os.getenv('LOGTO')
if verboseThreshold == nil then
    verboseThreshold = 0
else
    verboseThreshold = tonumber(verboseThreshold)
end

local term = {
    normal = '\027[0m',
    bright = '\027[1m',
    invert = '\027[7m',
    black = '\027[30m', 
    red = '\027[31m', 
    green = '\027[32m', 
    yellow = '\027[33m', 
    blue = '\027[34m', 
    magenta = '\027[35m',
    cyan = '\027[36m',
    white = '\027[37m', 
    default = '\027[39m'
}
Logger.type = {
    INFO = 0,
    WARNING = 1,
    ERROR = 2,
    FATAL = 3
}

logger = {}

log = nil

function logger.get(filename)
    if log == nil then
        log = Logger(filename)
    end
    return log
end

-------------------------------------------------------------------------------
function Logger:__init(filename)
    time = os.time()
    now =  os.date('*t', time)
    if filename ~= nil then
        self.filename = ('%s-%4d%2d%2d-%2d%2d%2d.log'):format(
            filename,
            now.year, now.month, now.day,
            now.hour, now.min, now.sec)
        dirname = path.dirname(filename)
        -- Create log file.
        local f = assert(io.open(filename, 'w'))
        f:close()
    else
        self.filename = nil
    end
end

-------------------------------------------------------------------------------
function Logger.typeString(typ)
    if typ == Logger.type.INFO then
        return ('%sINFO:%s'):format(term.green, term.default)
    elseif typ == Logger.type.WARNING then
        return ('%sWARNING:%s'):format(term.yellow, term.default)
    elseif typ == Logger.type.ERROR then
        return ('%sERROR:%s'):format(term.red, term.default)
    elseif typ == Logger.type.FATAL then
        return ('%sFATAL:%s'):format(term.red, term.default)
    else
        return 'UNKNOWN'
    end
end

-------------------------------------------------------------------------------
function Logger.typeStringPlain(typ)
    if typ == Logger.type.INFO then
        return 'INFO:'
    elseif typ == Logger.type.WARNING then
        return 'WARNING:'
    elseif typ == Logger.type.ERROR then
        return 'ERROR:'
    elseif typ == Logger.type.FATAL then
        return 'FATAL:'
    else
        return 'UNKNOWN:'
    end
end

-------------------------------------------------------------------------------
function Logger.timeString(time)
    local date = os.date('*t', time)
    return ('%04d-%02d-%02d %02d:%02d:%02d'):format(
        date.year, date.month, date.day, 
        date.hour, date.min, date.sec)
end

-------------------------------------------------------------------------------
function string.endswith(String, End)
    return End == '' or string.sub(String,-string.len(End)) == End
end

-------------------------------------------------------------------------------
function string.startswith(String, Start)
    return Start == '' or string.sub(String, 1, #Start) == Start
end

-------------------------------------------------------------------------------
function Logger:log(typ, text, verboseLevel)
    if verboseLevel == nil then
        verboseLevel = 0
    end
    if verboseLevel <= verboseThreshold then
        local info
        for i = 2,5 do
            info = debug.getinfo(i)
            if not string.endswith(info.short_src, 'logger.lua') then
                break
            end
        end
        local src = info.short_src
        if string.startswith(src, './') then
            src = string.sub(src, 3)
        end
        print(('%s %s %s:%d %s'):format(
            self.typeString(typ),
            self.timeString(os.time()),
            src, info.currentline, text))
    end
    if self.filename then
        local f = assert(io.open(filename, 'a'))
        f:write(('%s %s %s:%d %s'):format(
            self.typeString(typ),
            self.timeString(os.time()),
            src, info.currentline, text))
        f:close()
    end
end

-------------------------------------------------------------------------------
function Logger:logInfo(text, verboseLevel)
    if verboseLevel == nil then
        verboseLevel = 0
    end
    self:log(Logger.type.INFO, text, verboseLevel)
end

-------------------------------------------------------------------------------
function Logger:info(text, verboseLevel)
    self:logInfo(text, verboseLevel)
end

-------------------------------------------------------------------------------
function Logger:logWarning(text)
    self:log(Logger.type.WARNING, text, 0)
end

-------------------------------------------------------------------------------
function Logger:warning(text, verboseLevel)
    self:logWarning(text, verboseLevel)
end

-------------------------------------------------------------------------------
function Logger:logError(text)
    self:log(Logger.type.ERROR, text, 0)
end

-------------------------------------------------------------------------------
function Logger:error(text)
    self:logError(text)
end

-------------------------------------------------------------------------------
function Logger:logFatal(text)
    self:log(Logger.type.FATAL, text, 0)
    os.exit(0)
end

-------------------------------------------------------------------------------
function Logger:fatal(text)
    self:logFatal(text)
end

-------------------------------------------------------------------------------
return logger
